### PR TITLE
R2dbc provider

### PR DESCRIPTION
**Pull request by sokomishalov**
Draft for #700.
Also, I think it makes sense to provide an alternative reactive core API and implement it for providers which use reactive API (such as Lettuce, Redisson, Mongo, Cassandra). I might try to do it if it's ok for you from an API design perspective.

### CODE REVIEW

File: providers/r2dbc/shedlock-provider-r2dbc/pom.xml [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+    <build>
+        <plugins>
+            <plugin>
+                <groupId>org.apache.maven.plugins</groupId>
+                <artifactId>maven-jar-plugin</artifactId>
+                <configuration>
+                    <archive>
+                        <manifestEntries>
+                            <Automatic-Module-Name>
+                                net.javacrumbs.shedlock.provider.jdbc
```
Comment (lukas-krecan) [inline providers/r2dbc/shedlock-provider-r2dbc/pom.xml:77]: Should be changed

**Review by lukas-krecan — COMMENTED** [review]
Hi, first of all thanks a lot.

1) I do not want to provide an alterantive reactive core API. First of all, I can't imagine how it would help the default Spring or Micronaut implementation. Moreover, once Project Loom is out, people will be removing the ractive layer so I do not want to invest into reactive much.
2) I assume that the only added value is the fact that you already have R2DBC connection pool and you want to use it for ShedLock as well, is that correct? (You don't want to create synchronous JDBC DataSource just for ShedLock)
3) Regarding the implementation. I prefer [JdbcTemplateProvider](https://github.com/lukas-krecan/ShedLock/blob/master/providers/jdbc/shedlock-provider-jdbc-template/src/main/java/net/javacrumbs/shedlock/provider/jdbctemplate/JdbcTemplateStorageAccessor.java) since it provides more functionality (using DB time, using upserts etc.) The trouble is, that it is maintainable only thanks to named SQL parameters. I guess that R2DBC does not support named parameters or does it? If not, then support for DB specific code would be too complex, so your approach is ok.

File: providers/r2dbc/shedlock-provider-r2dbc/pom.xml (hunk 1 above)
Comment (sokomishalov) [reply providers/r2dbc/shedlock-provider-r2dbc/pom.xml]: my bad, fixed

**Comment by sokomishalov** [comment]
Thanks for the quick reply and review!
To the points:
1) It would help to use cluster locks in framework-independent reactive-streams pipelines. In spring I suppose we could rework aspects to deal with reactive publishers (just like `@Timed` annotation supports `Mono` and `Flux` now). 
Regarding Loom your point makes sense but is still arguable though. Loom has not been shifted to the latest LTS version (17) and I would not rely on it too much.
2) In my exact case - yes, I would like to reuse the existing r2dbc connection pool.
3) Yes, most r2dbc drivers don't support named parameters yet.

File: providers/r2dbc/shedlock-provider-r2dbc-internal/src/main/java/net/javacrumbs/shedlock/provider/r2dbc/internal/AbstractR2dbcStorageAccessor.java [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+import java.time.Instant;
+import java.util.function.BiFunction;
+import java.util.function.Function;
+
+import static java.util.Objects.requireNonNull;
+
+/**
+ * Internal class, please do not use.
+ */
+public abstract class AbstractR2dbcStorageAccessor extends AbstractStorageAccessor {
```
Comment (lukas-krecan) [inline providers/r2dbc/shedlock-provider-r2dbc-internal/src/main/java/net/javacrumbs/shedlock/provider/r2dbc/internal/AbstractR2dbcStorageAccessor.java:37]: What's thereason for extracting it to a new module? I would prefer to keep it package-protected inside the r2dbc module until it's needed elsewhere.

File: providers/r2dbc/shedlock-provider-r2dbc/src/main/java/net/javacrumbs/shedlock/provider/r2dbc/R2dbcStorageAccessor.java [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+        this.connectionFactory = requireNonNull(connectionFactory, "dataSource can not be null");
+    }
+
+    @Override
+    protected <T> Mono<T> executeCommand(
+        String sql,
+        Function<Statement, Mono<T>> body,
+        BiFunction<String, Throwable, Mono<T>> exceptionHandler
+    ) {
+        return Mono.usingWhen(
```
Comment (lukas-krecan) [inline providers/r2dbc/shedlock-provider-r2dbc/src/main/java/net/javacrumbs/shedlock/provider/r2dbc/R2dbcStorageAccessor.java:44]: Wow, that's why I really hope that Project Loom will be a thing. How is anyone supposed to understand code like this? Just ranting, it's not your fault.

File: providers/r2dbc/shedlock-provider-r2dbc-internal/src/main/java/net/javacrumbs/shedlock/provider/r2dbc/internal/AbstractR2dbcStorageAccessor.java (hunk 2 above)
Comment (sokomishalov) [reply providers/r2dbc/shedlock-provider-r2dbc-internal/src/main/java/net/javacrumbs/shedlock/provider/r2dbc/internal/AbstractR2dbcStorageAccessor.java:37]: I did it for the same reason as jdbc did - to provide the availability to implement lock provider via spring-r2dbc wrapper in another module in future.

File: providers/r2dbc/shedlock-provider-r2dbc/src/main/java/net/javacrumbs/shedlock/provider/r2dbc/R2dbcStorageAccessor.java (hunk 3 above)
Comment (sokomishalov) [reply providers/r2dbc/shedlock-provider-r2dbc/src/main/java/net/javacrumbs/shedlock/provider/r2dbc/R2dbcStorageAccessor.java:44]: It's a matter of habit actually. 
And yes, in some cases reactive and FP libraries reinvent a wheel with such constructions (in this case try-with-resources).

**Pull request merged**